¡Hola, Adolfo!

Mi nombre es Tonatiuh Cruz. Me complace revisar tu proyecto hoy.

Al identificar cualquier error inicialmente, simplemente los destacaré. Te animo a localizar y abordar los problemas de forma independiente como parte de tu preparación para un rol como data-scientist. En un entorno profesional, tu líder de equipo seguiría un enfoque similar. Si encuentras la tarea desafiante, proporcionaré una pista más específica en la próxima iteración.

Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o elimines**.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Éxito. Todo está hecho correctamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con comentarios en rojo.
</div>

Puedes responderme utilizando esto:

<div class="alert alert-block alert-info">


<div class="alert alert-block alert-success">
<b>Review General iteración 1) </b> <a class="tocSkip"></a>
    
Adolfo, has hecho un excelente trabajo con el desarrollo del proyecto, cada vez más cercas de convertirte en un cientifico de datos. Realizaste la carga de bases, su análisis inicial,el desarrollo de los modelos, el calculo del RMSE además lo replicate para todas la regiones y concluiste con el análisis de ganancias. Solamente te dejo algunos comentarios para complementar el análisis.

Sigue con el excelente trabajo!

# Proyecto: Selección de los mejores lugares de pozos petroleros para OilyGiant

## Introducción

La compañia de extracción de petróleo OilyGiant necesita encontrar **los mejores 200 lugares** para abrir pozos petroleros.

Tenemos que crear un modelo para predecir el volumen de reservas en los pozos nuevos,  de los cuáles se podrán elegir los que tengan valores más altos y con ésto la región con el beneficio total más alto para los pozos seleccionados.


### 1. Preparación de datos

In [27]:
# Importación de librerías
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
from scipy import stats as st

# Cargar y explorar datos
geo_data_0 = pd.read_csv('/datasets/geo_data_0.csv')
geo_data_1 = pd.read_csv('/datasets/geo_data_1.csv')
geo_data_2 = pd.read_csv('/datasets/geo_data_2.csv')

print('Región 0:')
print(geo_data_0.info())
display(geo_data_0.head())
print()
print('Región 1:')
print(geo_data_1.info())
display(geo_data_1.head())
print()
print('Región 2:')
print(geo_data_2.info())
display(geo_data_2.head())

Región 0:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647


Región 1:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305


Región 2:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

Después de cargar los datos y explorarlos, podemos ver que los datasets estan limpios y podemos trabajar con ellos, no tienen valores nulos y el tipo de datos es correcto para cada columna.

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Excelente trabajo con la carga de la información y su exploración inicial.  Solamente te recomiendo un análisis de los registros duplicados que como aprendiste en cursos pasados sirve para asegurar que nuestros resultados no estarán sesgados

In [28]:
# Revisamos si hay duplicados
print('Duplicados en dataset Región 0:', geo_data_0.duplicated().sum())
print('Duplicados en dataset Región 1:', geo_data_1.duplicated().sum())
print('Duplicados en dataset Región 2:', geo_data_2.duplicated().sum())

Duplicados en dataset Región 0: 0
Duplicados en dataset Región 1: 0
Duplicados en dataset Región 2: 0


<div class="alert alert-block alert-info">
    Tienes razón, solamente verifique que no hubiera valores nulos y asumí que no había valores duplicados y que todos los datos estaban limpios. Ahora si ya verificando, puedo decir que los datasets no cuentan con valores duplicados y podemos trabajar con ellos. 


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Gran trabajo con el ajuste!

### 2. Entrenamiento y validación de modelo

Se va a entrenar un modelo y se va a probar para cada una de las regiones.

In [29]:

# Como se va a realizar lo mismo para cada región creamos una función:



def train_validate_model(data):
    features = data[['f0', 'f1', 'f2']]
    target = data['product']
    X_train, X_valid, y_train, y_valid = train_test_split(features, target, test_size=0.25, random_state=12345)
    
    model = LinearRegression()
    model.fit(X_train, y_train)
    predictions = model.predict(X_valid)
    rmse = mean_squared_error(y_valid, predictions, squared=False)
    
    print(f"RMSE: {rmse:.2f}")
    print(f"Media predicha: {predictions.mean():.2f}")

    
    return model, predictions, y_valid




<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien, dividiste adecuadamente los datos en entrenamiento, validación. Los dos conjuntos son muy importantes, el de validación en particular nos es útil para probar los diferentes hiperparámetros antes de hacer el testeo final con el set de prueba.
</div>

Entrenamos ahora para las tres regiones:


In [30]:

print('Resultados para región 0:')
model_0, predictions_0, y_valid_0 = train_validate_model(geo_data_0)


Resultados para región 0:
RMSE: 37.58
Media predicha: 92.59


In [31]:

print('Resultados para región 1:')
model_1, predictions_1, y_valid_1 = train_validate_model(geo_data_1)


Resultados para región 1:
RMSE: 0.89
Media predicha: 68.73


In [32]:

print('Resultados para región 2:')
model_2, predictions_2, y_valid_2 = train_validate_model(geo_data_2)


Resultados para región 2:
RMSE: 40.03
Media predicha: 94.97


Realizamos un **análisis de los resultados**:

La región 1 tiene un RMSE muy bajo, lo que arroja que es el modelo que tiene una precisión muy alta, al contrario que las regiones 0 y 2, siendo asi que el modelo de la región 1 sería el más confiable.

Ahora bien, por cuanto a la media, las regiones 0 y 2, muestran el mayor volumen de reservas estimadas. Sin embargo, en estas regiones no hay tanta precisión en las predicciones


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Adolfo, implementaste el modelo de forma excelente para los 3 conjuntos de datos! En este punto ya puedes obervar cuál es el modelo con el mayor $R^2$ y menor RMSE. 
    
Solamente recuerda complementarlo con el escalado de los datos. El escalado es muy útil para mejorar el rendimiento del modelo al asegurar que todas las características tienen el mismo rango.     
</div>

<div class="alert alert-block alert-info">
Importé al principio el módulo de StandardScaler y realizo a continuación la adaptación de la función con el escalado de los datos.


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Gran trabajo con el ajuste!

In [35]:
def train_validate_model_scaled(data):
    features = data[['f0', 'f1', 'f2']]
    target = data['product']
    
    X_train, X_valid, y_train, y_valid = train_test_split(features, target, test_size=0.25, random_state=12345)
    
    # Escalado de características
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_valid_scaled = scaler.transform(X_valid)
    
    # Entrenamiento del modelo
    model = LinearRegression()
    model.fit(X_train_scaled, y_train)
    predictions = model.predict(X_valid_scaled)
    rmse = mean_squared_error(y_valid, predictions, squared=False)
    
    print(f"RMSE: {rmse:.2f}")
    print(f"Media predicha: {predictions.mean():.2f}")
    
    return model, predictions, y_valid, scaler

Ahora lo aplico a las tres regiones:

In [36]:
print('Resultados para región 0:')
model_0, predictions_0, y_valid_0, scaler_0 = train_validate_model_scaled(geo_data_0)

print('Resultados para región 1:')
model_1, predictions_1, y_valid_1, scaler_1 = train_validate_model_scaled(geo_data_1)

print('Resultados para región 2:')
model_2, predictions_2, y_valid_2, scaler_2 = train_validate_model_scaled(geo_data_2)

Resultados para región 0:
RMSE: 37.58
Media predicha: 92.59
Resultados para región 1:
RMSE: 0.89
Media predicha: 68.73
Resultados para región 2:
RMSE: 40.03
Media predicha: 94.97


<div class="alert alert-block alert-info">
Por suerte veo que no cambiaron los resultados, pero estaré más consciente de aplicar el escalado de datos.

### 3. Cálculo de ganancias.

In [21]:

# Parámetros del negocio
BUDGET = 100_000_000
NUM_WELLS = 200
COST_PER_WELL = BUDGET / NUM_WELLS
REVENUE_PER_UNIT = 4500
MIN_UNITS_FOR_PROFIT = COST_PER_WELL / REVENUE_PER_UNIT




### 4. Función para calcular ganancias.

Obtenemos los valores más altos para poder tener las 200 predicciones más altas, luego vamos a sumar el total para estos 200 pozos y luego lo multiplicaremos por el ingreso por unidad.

In [22]:

# Función para calcular ganancias


def calculate_profit(predictions, actuals):
    # Obtener los índices de las 200 predicciones más altas
    top_200_indices = np.argsort(predictions)[-200:]
    
    # Seleccionar los valores reales y predichos correspondientes
    top_200_actuals = actuals.iloc[top_200_indices]
    top_200_predictions = predictions[top_200_indices]
    
    # Calcular ganancias
    total_units = top_200_actuals.sum()
    total_predicts = top_200_predictions.sum()
    revenue = total_units * REVENUE_PER_UNIT
    print(f"Ganancia estimada: ${revenue:,.2f}")
    print(f"Volumen de reservas predichas: {total_predicts}")
    
    # Devolver también las predicciones seleccionadas
    return revenue, total_units, top_200_predictions





Ahora lo aplicamos a cada una de las regiones y se almacenan las predicciones (top_200_predictions) para cada una de las 3 regiones.

In [23]:
print('Región 0')
revenue_0, units_0, top_preds_0 = calculate_profit(predictions_0, y_valid_0)

Región 0
Ganancia estimada: $133,208,260.43
Volumen de reservas predichas: 31102.330838811402


In [24]:
print('Región 1')
revenue_1, units_1, top_preds_1 = calculate_profit(predictions_1, y_valid_1)

Región 1
Ganancia estimada: $124,150,866.97
Volumen de reservas predichas: 27746.026782163433


In [25]:
print('Región 2')
revenue_2, units_2, top_preds_2 = calculate_profit(predictions_2, y_valid_2)

Región 2
Ganancia estimada: $127,103,499.64
Volumen de reservas predichas: 29603.89865831835


**conclusones** de los resultados obtenidos:

Podemos ver que todas las regiones generán ganancias conforme al monto de inversión, de esta manera la **región 0** tiene la **mayor ganancia** estimada y el **mayor volumen de reservas predichas**, lo que la convierte en la candidata más fuerte hasta ahora.

La diferencia entre la región 0 y la región 2 es de más de $6 millones, lo cual es significativo en términos de inversión. Mi propuesta a esta altura del análisis sería la región 0 para el desarrollo de los pozos petrolíferos.

### 5. Riesgos y ganancias por región.

In [11]:
state = np.random.RandomState(12345)

In [12]:
# Realizamos una función para despues poder aplicarla a cada región.

def bootstrap_profit(predictions, actuals, n_bootstrap=1000, num_wells=NUM_WELLS, revenue_per_unit=REVENUE_PER_UNIT, budget=BUDGET):
    profits = []

    for _ in range(n_bootstrap):
        # Selección aleatoria de 500 pozos como muestra
        indices = state.choice(len(predictions), size=500, replace=False)
        sample_predictions = predictions[indices]
        sample_actuals = actuals.iloc[indices].reset_index(drop=True)

        # Selección de los 200 pozos con mayor predicción
        top_indices = sample_predictions.argsort()[-num_wells:]
        selected_actuals = sample_actuals[top_indices]

        # Cálculo de ganancia
        total_revenue = selected_actuals.sum() * revenue_per_unit
        profit = total_revenue - budget
        profits.append(profit)

    profits = pd.Series(profits)
    mean_profit = profits.mean()
    conf_interval = (profits.quantile(0.025), profits.quantile(0.975))
    loss_risk = (profits < 0).mean() * 100

    return mean_profit, conf_interval, loss_risk


    

y ahora lo aplicamos a cada una de las regiones

In [13]:
print("Región 0:")
mean_0, conf_0, risk_0 = bootstrap_profit(predictions_0, y_valid_0)
print(f"Ganancia promedio: ${mean_0:,.2f}")
print(f"Intervalo de confianza 95%: (${conf_0[0]:,.2f}, ${conf_0[1]:,.2f})")
print(f"Riesgo de pérdida: {risk_0:.2f}%\n")


Región 0:
Ganancia promedio: $3,807,108.91
Intervalo de confianza 95%: ($-1,269,476.38, $8,796,139.68)
Riesgo de pérdida: 7.20%



In [14]:
print("Región 1:")
mean_1, conf_1, risk_1 = bootstrap_profit(predictions_1, y_valid_1)
print(f"Ganancia promedio: ${mean_1:,.2f}")
print(f"Intervalo de confianza 95%: (${conf_1[0]:,.2f}, ${conf_1[1]:,.2f})")
print(f"Riesgo de pérdida: {risk_1:.2f}%\n")



Región 1:
Ganancia promedio: $4,547,854.35
Intervalo de confianza 95%: ($467,300.85, $8,402,133.56)
Riesgo de pérdida: 1.30%



In [15]:
print("Región 2:")
mean_2, conf_2, risk_2 = bootstrap_profit(predictions_2, y_valid_2)
print(f"Ganancia promedio: ${mean_2:,.2f}")
print(f"Intervalo de confianza 95%: (${conf_2[0]:,.2f}, ${conf_2[1]:,.2f})")
print(f"Riesgo de pérdida: {risk_2:.2f}%")


Región 2:
Ganancia promedio: $3,892,170.74
Intervalo de confianza 95%: ($-1,156,095.66, $9,065,125.90)
Riesgo de pérdida: 7.30%


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo con el calculo de las ganacias para cada una de las regiones. Posteriormente este resultados lo complementas con el análisis de riesgos


## Conclusión y propuesta

Aunque la Región 0 tuvo la mayor ganancia estimada en el paso 4, el análisis de riesgo revela que Región 1 es la única que cumple con el criterio de riesgo de pérdida inferior al 2.5%. Además, su ganancia promedio es la más alta entre las regiones que cumplen con ese umbral.
Por lo tanto, la **Región 1** es la **opción óptima** para el desarrollo de los 200 pozos petrolíferos, ya que ofrece:
- Ganancia promedio más elevada.
- Riesgo de pérdida significativamente bajo.
- Intervalo de confianza completamente positivo (sin valores negativos en el extremo inferior).
